text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+23d9244c918044b4f1e28de295065b41e029f37c commit 1122 848 12
+ff51a7fc40b7fb6f165d3a92ab88d8d2bf8c0b18 commit 1117 844 860
+...
+```
+
+Pack files are archives that contain multiple objects. They make things a little easier on the filesystem and the HTTP transfers, but more importantly delta compression is applied when files are in a pack.
+
+### Git summary
+
+Git stores different types of objects indexed by hashes, compressed in a very simple key-value scheme on the local filesystem. Every object contains a header indicating its type (commit, tree, blob). These objects can be fetched by hash even across the network. Parent relationships are indicated by hash as well.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+* What is the actual impact on the execution engine?
+* Do we serialize hashes or contents?
+* Do we store DAG nodes statically or dynamically (`Box<dyn ...>`)?
+
+All these questions should be answered before this RFC is merged.
+
+## Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+While not covered in detail in this RFC, these changes lay the groundwork for exciting future possiblities:
+
+### CDNs
+
+Since retrieving objects is just a matter of fetching by hash, we can leverage a CDN, for example via HTTP. Whenever a node needs an object `435abcf`, it can try to `GET http://casper.cdn.net/objects/435abcf` first. If found, we just saved the network the burden of sending the missing piece. This is not limited to the "official" CDN, validators are free to run one themselves, we may even add mechanisms to discover them.
+
+Open questions with this approach is who puts data on the CDN in the first place:
+
+* An easy first take is to upload daily snapshots of the whole chain ourselves. This already considerably reduces the burden on the network when new nodes are joining.
+* Another approach is to allow any node to push objects to the CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus specialized software.
+* Clients could also upload WASM blocks to a CDN beforehand, making the actual deploy really small.
+
+Note that none of this is strictly required as all the CDN functionality is strictly optional with regards to the correct operation of the network.
+
+### Git-style pack files
+
+We may copy the idea of pack files, saving us some roundtrips. This is the first more complicated feature, but a node joining could easily happen to find a pack file that contains the
+blocks (without deploy contents) all the way up to genesis. It will then fetch other information as required.
+
+A pack is just a collection of objects, so importing one into a node should be (disk space not withstanding) strictly a benefit to the node. In the absence of CDNs, pack files can be used to speed up inter-node downloads.
+
+It is entirely possible to implement pack files as just a serialization of `HashMap<ObjectHash, Object>`.
```

Comment:
Yes, Pack files solve the "many small files" issues on the network, not locally. We want to avoid making a lot of requests though, instead sending small bits in bulk.

It also allows for easy shipping of archives.